### PR TITLE
Add shared memory region with DAX support to virtio-fs

### PR DESCRIPTION
**Pull request by sboeuf**
This pull request enhances the current version of the virtio-fs device. It introduces a new shared memory region with the guest. On behalf of the vhost-user deamon, the VMM can now mmap the shared files through this shared region, so that we can achieve very good read/write performances when using virtio-fs.

### CODE REVIEW

**Comment by sboeuf** [comment]
@rbradford @sameo 
The PR is now ready to be reviewed.

Here is a list of things left to do, but they might be addressed in follow up PRs:
TODO:
- [ ] Update the kernel to the branch 5.3-rcX from virtio-fs maintainers. Relying on this branch, the `virtio-pmem` patches will already be integrated since they are part of the upstream kernel. Using new kernel means we need to update the way the virtio-fs mount is performed inside the guest (syntax changed)
- [ ] Update `virtiofsd`. We need to get the `virtiofsd` binary from somewhere else than NEMU since we need the latest patches (branch virtio-fs-dev) on the virtio-fs/qemu.git to properly support virtio-fs. Additionally, we need one patch that I submitted to qemu-devel to fix libvhost-user, which will hopefully be merged.
- [ ] Add some integration tests using the new `cache_size` option from `--fs` and combine it with `virtiofsd` being run with the option `cache=always` instead of `cache=none`

File: vmm/src/config.rs [hunk 1]
```diff
@@ -281,11 +283,14 @@ impl<'a> FsConfig<'a> {
...
             } else if param.starts_with("queue_size=") {
                 queue_size_str = &param[11..];
+            } else if param.starts_with("cache_size=") {
+                cache_size_str = &param[11..];
             }
         }
 
         let mut num_queues: usize = 1;
         let mut queue_size: u16 = 1024;
+        let mut cache_size: Option<u64> = None;
```
Comment (rbradford) [inline vmm/src/config.rs]: No default value?

File: vmm/src/vm.rs [hunk 2]
```diff
@@ -820,16 +831,69 @@ impl DeviceManager {
...
+                        let fs_guest_addr = allocator
+                            .allocate_mmio_addresses(
+                                None,
+                                fs_cache as GuestUsize,
+                                Some(0x0020_0000),
+                            )
+                            .ok_or(DeviceManagerError::FsRangeAllocation)?;
+
+                        let addr = unsafe {
+                            libc::mmap(
```
Comment (rbradford) [inline vmm/src/vm.rs]: Probably should look at errno.

File: vm-virtio/src/fs.rs [hunk 3]
```diff
@@ -85,16 +91,74 @@ pub enum Error {
...
 type Result<T> = result::Result<T, Error>;
 
-struct FsEpollHandler {
+struct SlaveReqHandler {
+    mmap_cache_addr: u64,
+}
+
+impl VhostUserMasterReqHandler for SlaveReqHandler {
+    fn handle_config_change(&mut self) -> HandlerResult<()> {
+        println!("handle_config_change");
```
Comment (rbradford) [inline vm-virtio/src/fs.rs]: Residual debugging

File: vm-virtio/src/fs.rs [hunk 4]
```diff
@@ -85,16 +91,74 @@ pub enum Error {
...
+        Ok(())
+    }
+
+    fn fs_slave_map(&mut self, fs: &VhostUserFSSlaveMsg, fd: RawFd) -> HandlerResult<()> {
+        println!("fs_slave_map");
```
Comment (rbradford) [inline vm-virtio/src/fs.rs]: Ditto

File: vmm/src/vm.rs (hunk 2 above)
Comment (sboeuf) [reply vmm/src/vm.rs]: Yes I forgot to check the error returned here, I'm gonna fix that.

File: vmm/src/config.rs (hunk 1 above)
Comment (rbradford) [reply vmm/src/config.rs]: So using this command line, that has no mention about DAX, radically changes the behaviour? I'm not sure about that...

**Comment by sboeuf** [comment]
@rbradford @sameo 
I have addressed all the points from the TODO list that were missing, either through this PR or through #170 
Please have a look!

**Comment by sameo** [comment]
Except for a adbafba3da7bc493eada38729ee155c21c66045e commit message micro nit(We're removing the pci, not virtio-pci dependency), this looks good to me.

Out of curiosity, have we measured the actual improvements?

**Comment by sboeuf** [comment]
@rbradford 
>Looks good, we can always address my suggestion #170 about changing the variable name in #69 anyway so this could just be merged as is.

Cool! Since the PR is merged, let's hope @bjzhjing will make the change in #69 

@sameo 
>Except for a adbafba commit message micro nit (We're removing the pci, not virtio-pci dependency), this looks good to me.

Oh yes you're right. Sorry I missed that.

>Out of curiosity, have we measured the actual improvements?

No I haven't but that would be very nice to do indeed. I'll ask virtio-fs maintainers about the best way to get performances tested.

**Pull request merged**